gravitational/teleport
# Surface target ports in tsh and Connect

* Part of #46758.
* Depends on:
  * #49453
  * #49505
  * #49625

This PR makes it so that both `tsh apps ls` and Connect are capable of showing target ports supported by TCP apps. See [the UI section of the RFD](https://github.com/gravitational/teleport/blob/master/rfd/0182-multi-port-tcp-app-access.md#ui).

It also makes sure that any VNet modals or notifications in Connect include the target port where possible.

```
$ tsh apps ls
Application      Description Type Public Address                      Target Ports    Labels
---------------- ----------- ---- ----------------------------------- --------------- --------
dumper                       HTTP dumper.teleport-local.dev
example1         Example app HTTP example1.teleport-local.dev                         env=test
simplehttpserver             HTTP simplehttpserver.teleport-local.dev
tcp-postgres                 TCP  tcp-postgres.teleport-local.dev
waldo                        TCP  waldo.teleport-local.dev            4080-4090, 5095 env=prod
```
<img alt="Connect available ports" src="https://github.com/user-attachments/assets/4580a34f-fcf2-4613-b9c2-d9e64d19f00f" width="480">

## How to test

To test the display of target ports, you need to compile teleport from this branch and then add any bogus multi-port TCP app – the ports don't have to work.

```yaml
    - name: multi-port-example
      uri: tcp://localhost
      tcp_ports:
        - port: 5434 # Postgres
        - port: 8765 # HTTP echo
        - port: 63636
          end_port: 64321
```

To test the notification and the modal, you can set up any kind of HTTP app on 8765, start VNet and then try to curl the app at `multi-port-example.<your cluster>:8765`. Just make sure the address of the app is not in your `/etc/hosts`. Turning on per-session MFA and then closing its modal will let you test the notification. The modal can be seen in the story or by first curling the app, waiting for the cert to expire and then attempting to curl it again – it's important that you curl it at least once before the cert expires.
